Mathlib/Tactic/Linter/UnusedAssumptionInType.lean

```diff
@@ -0,0 +1,115 @@
+/-
+Copyright (c) 2024 Yury Kudryashov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov, Michael Rothgang
+-/
+import Batteries.Tactic.Lint
+import Mathlib.Lean.Expr.Basic
+
+/-!
+# Linters for unused assumptions in a type
+
+This file defines a collection of similar linters, which check for declarations assuming a
+hypothesis of the form `[Type p]` which contains an assumption about `p` that is not used:
+this includes `Decidable p`, `Encodable p`, `Inhabited p` and others.
+Usually, the code can be restructured to avoid these assumptions, making the statements in question
+more general.
+-/
+
+namespace Batteries.Tactic.Lint
+open Lean Meta
+
+/-- Check if a given declaration assumes some hypothesis `[Type p]`, but doesn't use this
+assumption in the type. `typesToAvoid` the list of such types.
+This is the main logic underlying the linters below. -/
+def checkUnusedAssumptionInType (declInfo : ConstantInfo) (typesToAvoid : Array Name) :
+    MetaM (Option MessageData) := do
+  -- We omit inductive types and their constructors, to reduce false positives.
+  -- We also omit partial declarations (via the `opaque` definitions they generate):
+  -- these are less useful for theorem proving, hence the linter is less useful there.
+  if declInfo.isInductive || declInfo.isCtor ||
+      (declInfo matches .opaqueInfo ..) then return none
```

Comment:
It's not a performance improvement, but you might like this syntax:
```lean4
if declInfo matches .inductInfo .. | .ctorInfo .. | .opaqueInfo .. then return none
```

Reply:
I actually know this syntax - but thanks for reminding me! This started as one condition, branched out and I didn't re-visit: I will certainly use the or pattern!